Intel OpenVINO backend

dkurt opened the pull request:
To enable OpenVINO, wrap a model by `ml3d.models.OpenVINOModel`

```python
net = ml3d.models.PointPillars(**cfg.model, device='cpu')
ov_net = ml3d.models.OpenVINOModel(net)
```

## Enabled models

| Model / Framework    | ported to OpenVINO |
|--------------------|---------------|
| RandLA-Net (tf)    |  + |
| RandLA-Net (torch) | + (Gather takes too much memory and execution time)
| KPConv     (tf)    | + |
| KPConv     (torch) | + |
| SparseConvUnet (torch)|
| SparseConvUnet (tf)| 
| PointPillars (tf)    | |
| PointPillars (torch) | + |
| PointRCNN (tf)       | |

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/isl-org/open3d-ml/406)
<!-- Reviewable:end -->

dkurt commented:
Not sure why tests are red (https://github.com/dkurt/Open3D-ML/runs/4212507141?check_suite_focus=true). Can you please guide me here?

<details>
<summary>logs</summary>

```
============================= test session starts ==============================
platform linux -- Python 3.6.15, pytest-6.0.1, py-1.11.0, pluggy-0.13.1
Using --randomly-seed=1825362725
rootdir: /home/runner/work/Open3D-ML/test_workdir
plugins: randomly-3.8.0
collected 8 items

tests/test_models.py ......                                              [ 75%]
tests/test_integration.py ..                                             [100%]

=============================== warnings summary ===============================
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/pkg_resources/_vendor/packaging/version.py:114: 616 warnings
  /opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/pkg_resources/_vendor/packaging/version.py:114: DeprecationWarning: Creating a LegacyVersion has been deprecated and will be removed in the next major release
    DeprecationWarning,

/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/tensorflow/python/autograph/impl/api.py:22
  /opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/tensorflow/python/autograph/impl/api.py:22: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

tests/test_models.py::test_pointpillars_torch
tests/test_models.py::test_kpconv_torch
  /opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/torch/onnx/utils.py:355: UserWarning: Skipping _decide_input_format
   -1
    warnings.warn("Skipping _decide_input_format\n {}".format(e.args[0]))

tests/test_models.py::test_kpconv_torch
  /opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/torch/onnx/symbolic_opset9.py:662: UserWarning: This model contains a squeeze operation on dimension 2. If the model is intended to be used with dynamic input shapes, please use opset version 11 to export the model.
    "intended to be used with dynamic input shapes, please use opset version 11 to export the model.")

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================= 8 passed, 620 warnings in 324.95s (0:05:24) ==================
[Open3D INFO] Memory Statistics: (Device) (#Malloc) (#Free)
[Open3D INFO] ---------------------------------------------
[Open3D WARNING] CPU:0: 84 80 --> 4 with 144 total bytes
[Open3D WARNING]     0x56092a55f2c0 @ 36 bytes
[Open3D WARNING]     0x56091f779450 @ 36 bytes
[Open3D WARNING]     0x56092412a220 @ 36 bytes
[Open3D WARNING]     0x56093e938dc0 @ 36 bytes
[Open3D INFO] ---------------------------------------------
Error: Process completed with exit code 1.
```

</details>

sanskar107 commented:
> Not sure why tests are red (https://github.com/dkurt/Open3D-ML/runs/4212507141). Can you please guide me here?
> 
> logs

@dkurt This is probably caused by this (https://github.com/isl-org/Open3D/blob/master/cpp/open3d/core/MemoryManagerStatistic.cpp#L60).
Open3D has custom reference counter which tracks all the Tensor objects being created and destroyed. It gets complicated when torch Tensor is passed as a reference with shared memory. A way to resolve this is to strip down the code to a minimum changes and see if the error is still there.

dkurt commented:
@sanskar107, I've found that `torch.onnx.export` leads to memory leak. That's an internal Torch function.

dkurt commented:
@sanskar107, thanks! Solved by the latest commit. See tests: https://github.com/dkurt/Open3D-ML/runs/4308998309?check_suite_focus=true

sanskar107 commented:
> @sanskar107, thanks! Solved by the latest commit. See tests: https://github.com/dkurt/Open3D-ML/runs/4308998309?check_suite_focus=true

Great. Is this PR ready for review?

dkurt commented:
@sanskar107, yes, please. Thanks in advance!

dkurt commented:
Hi, @sanskar107! Just wanted to ask if you had a chance to review the proposal.

sanskar107 commented on ml3d/torch/models/kpconv.py:
```diff
@@ -1245,7 +1245,7 @@ def forward(self, x):
             x = self.batch_norm(x)
             x = x.transpose(0, 2)
-            return x.squeeze()
+            return x.squeeze(2)
```
Its a good fix

sanskar107 commented on requirements-tensorflow.txt:
```diff
@@ -1 +1,2 @@
 tensorflow~=2.5.2
+openvino-dev==2021.4.1
```
Why do we have a different name for tensorflow?

dkurt replied:
Package `openvino` is just a runtime and it can run ONNX models (PyTorch). For TensorFlow models it's required to use additional tool Model Optimizer which goes with `openvino-dev`. So `openvino-dev = MO + openvino`

sanskar107 commented on .github/workflows/ubuntu.yml:
```diff
@@ -3,8 +3,7 @@ name: Ubuntu CI
   push:
-    branches:
-      - master
+    branches: [master, dev]
```
@ssheorey do we want this addition?

dkurt replied:
Can revert it,  no problem. It's just for testing

ssheorey reviewed (changes requested):
Hi @dkurt thanks very much for bringing OpenVINO support to Open3D-ML! The PR looks good to me after a couple of changes:
- We want to ensure that OpenVino is not a hard dependency for running Open3D-ML. I have  a few suggestions below to ensure that. Also test that the CI runs locally without openvino installed.
- Add OpenVINO support to the documentation so that users know about it. Create a `docs/openvino.md` file with some examples of how to use the model with OpenVino. You can expand on your PR description. Include the support matrix from there. Also mention Intel HW support (I believe it is only CPU, not GPU, fpga, etc.). Add a link to `docs/openvino.md` from the main `readme.md` file.

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/406#-:-MrA7lk94pu_-P1oSAN3:bo0z09v)* status: 0 of 12 files reviewed, 14 unresolved discussions (waiting on @dkurt, @sanskar107, and @ssheorey)

___
*[requirements-tensorflow.txt, line 2 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MnPX47qmha3GTuSSmS_-r2-2:-Mr9zyEY4p3AEL-mJdvW:b-tcskzp) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/requirements-tensorflow.txt#L2)):*
<details><summary><i>Previously, dkurt (Dmitry Kurtaev) wrote…</i></summary><blockquote>

Package `openvino` is just a runtime and it can run ONNX models (PyTorch). For TensorFlow models it's required to use additional tool Model Optimizer which goes with `openvino-dev`. So `openvino-dev = MO + openvino`
</blockquote></details>

The latest OpenVino is `2021.4.2`. Is it OK to upgrade to that?
___
*[.github/workflows/ubuntu.yml, line 6 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MnPX47qmha3GTuSSmSg-r2-6:-MrA-c1LCNQsELezLTgi:bwljgte) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/.github/workflows/ubuntu.yml#L6)):*
<details><summary><i>Previously, dkurt (Dmitry Kurtaev) wrote…</i></summary><blockquote>

Can revert it,  no problem. It's just for testing
</blockquote></details>

Its a good idea to run CI for push to `dev` as well. Let's keep it.
___
*[ml3d/tf/models/__init__.py, line 10 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA2AoB26uvs4kHg2Xf:-MrA2AoB26uvs4kHg2Xg:b-pypro1) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/ml3d/tf/models/__init__.py#L10)):*
> ```Python
> from .point_transformer import PointTransformer
> from .pvcnn import PVCNN
> from .openvino_model import OpenVINOModel
> ```

We don't want OpenVino to be a hard dependency for using Open3D-ML. Can we put all OpenVino specific imports in a try-except block and add OpenVINOModel to `__all__` if import succeeds ? Also check that CI runs without OpenVINO installed on your local machine.
___
*[ml3d/tf/models/openvino_model.py, line 11 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA2fN37Seb4mbIPXHK:-MrA2fN37Seb4mbIPXHL:b9h4v8) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/ml3d/tf/models/openvino_model.py#L11)):*
> ```Python
> 
> 
> class OpenVINOModel:
> ```

docstring
___
*[ml3d/tf/models/openvino_ext/front/tf/min.py, line 6 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA2ujV5HB3T624Hfxj:-MrA2ujV5HB3T624Hfxk:bobnzze) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/ml3d/tf/models/openvino_ext/front/tf/min.py#L6)):*
> ```Python
> 
> 
> class MinFrontExtractor(FrontExtractorOp):
> ```

docstring. Why is this specific op needed?
[nit] Also, is the `tf` folder needed, since we are already in an upper level `tf` folder?
___
*[ml3d/torch/models/__init__.py, line 10 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA3DgpCYC90mpUdrWO:-MrA3Dgq0g0qXkc79BWO:b-pypro1) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/ml3d/torch/models/__init__.py#L10)):*
> ```Python
> from .point_transformer import PointTransformer
> from .pvcnn import PVCNN
> from .openvino_model import OpenVINOModel
> ```

We don't want OpenVino to be a hard dependency for using Open3D-ML. Can we put all OpenVino specific imports in a try-except block and add OpenVINOModel to `__all__` if import succeeds ? Also check that CI runs without OpenVINO installed on your local machine.
___
*[ml3d/torch/models/openvino_model.py, line 17 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA3PDm9KNBsV7lq6l6:-MrA3PDm9KNBsV7lq6l7:b9h4v8) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/ml3d/torch/models/openvino_model.py#L17)):*
> ```Python
> 
> 
> class OpenVINOModel:
> ```

docstring
___
*[ml3d/torch/models/openvino_model.py, line 58 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA3a3bEGkptQ3hDhLz:-MrA3a3bEGkptQ3hDhM-:b-cdxnft) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/ml3d/torch/models/openvino_model.py#L58)):*
> ```Python
>             }
>         elif not isinstance(inputs, dict):
>             raise Exception(f"Unknown inputs type: {inputs.__class__}")
> ```

`Exception` -> `TypeError`
___
*[tests/test_models.py, line 60 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA4ZTb2HBVViKtyNDo:-MrA4ZTb2HBVViKtyNDp:bhxsqru) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/tests/test_models.py#L60)):*
> ```Python
>     assert out.shape == (1, 5000, 10)
> 
>     # TODO: issue with Gather
> ```

Add explicit comment: Gather takes too much memory and execution time.
Make OpenVino checks optional by testing importing openvino and running the openvino checks if import was successful.
___
*[tests/test_models.py, line 100 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA7Z7t0lsVHCEgw91r:-MrA7Z7t0lsVHCEgw91s:bzqm9f) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/tests/test_models.py#L100)):*
> ```Python
>     assert out.shape == (1, 5000, 10)
> 
>     ov_net = ml3d.models.OpenVINOModel(net)
> ```

Make OpenVino checks optional by testing importing openvino and running the openvino checks if import was successful.
___
*[tests/test_models.py, line 136 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA7_t64tGbl_vmzLjL:-MrA7_t64tGbl_vmzLjM:bzqm9f) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/tests/test_models.py#L136)):*
> ```Python
>     assert out.shape[1] == 5
> 
>     ov_net = ml3d.models.OpenVINOModel(net)
> ```

Make OpenVino checks optional by testing importing openvino and running the openvino checks if import was successful.
___
*[tests/test_models.py, line 182 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA7ch56nKmOOdnS4Ny:-MrA7ch56nKmOOdnS4Nz:bzqm9f) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/tests/test_models.py#L182)):*
> ```Python
>     assert out.shape == (1000, 5)
> 
>     ov_net = ml3d.models.OpenVINOModel(net)
> ```

Make OpenVino checks optional by testing importing openvino and running the openvino checks if import was successful.
___
*[tests/test_models.py, line 213 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA7dRkECfHX7YQqLqs:-MrA7dRl5pl4dTTDPgdt:bzqm9f) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/tests/test_models.py#L213)):*
> ```Python
>         assert type(boxes) == list
> 
>     ov_net = ml3d.models.OpenVINOModel(net)
> ```

Make OpenVino checks optional by testing importing openvino and running the openvino checks if import was successful.


<!-- Sent from Reviewable.io -->

dkurt commented:
Hi, @ssheorey! Thanks for review!

> * We want to ensure that OpenVino is not a hard dependency for running Open3D-ML. I have  a few suggestions below to ensure that. Also test that the CI runs locally without openvino installed.

Added checks for OpenVINO availability so now `OpenVINOModel` class and corresponding tests are flexible. Should I 

> * Add OpenVINO support to the documentation so that users know about it. Create a `docs/openvino.md` file with some examples of how to use the model with OpenVino. You can expand on your PR description. Include the support matrix from there. Also mention Intel HW support (I believe it is only CPU, not GPU, fpga, etc.). Add a link to `docs/openvino.md` from the main `readme.md` file.

Done. Also added missed docstrings. Sorry about that.

> _[requirements-tensorflow.txt, line 2 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MnPX47qmha3GTuSSmS_-r2-2:-Mr9zyEY4p3AEL-mJdvW:b-tcskzp) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/requirements-tensorflow.txt#L2)):_
> 
> Previously, dkurt (Dmitry Kurtaev) wrote…
> The latest OpenVino is `2021.4.2`. Is it OK to upgrade to that?

Sure! Updated requirements.

> _[ml3d/tf/models/openvino_ext/front/tf/min.py, line 6 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA2ujV5HB3T624Hfxj:-MrA2ujV5HB3T624Hfxk:bobnzze) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/ml3d/tf/models/openvino_ext/front/tf/min.py#L6)):_
> 
> > ```python
> > class MinFrontExtractor(FrontExtractorOp):
> > ```
> 
> docstring. Why is this specific op needed? [nit] Also, is the `tf` folder needed, since we are already in an upper level `tf` folder?

This is a workaround for missed `Min` op from OpenVINO. It's fixed in on master branch and can be removed once next OpenVINO release is published. The structure of the folder is just something what OpenVINO expects (`front/tf/op.py`). Moved `optnvino_ext` to `tf/models/utils` to hide a bit.

> _[tests/test_models.py, line 60 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrA4ZTb2HBVViKtyNDo:-MrA4ZTb2HBVViKtyNDp:bhxsqru) ([raw file](https://github.com/isl-org/open3d-ml/blob/6311491760652647929f7ae1c9824a47792bf1d2/tests/test_models.py#L60)):_
> 
> > ```python
> >     assert out.shape == (1, 5000, 10)
> > 
> >     # TODO: issue with Gather
> > ```
> 
> Add explicit comment: Gather takes too much memory and execution time. Make OpenVino checks optional by testing importing openvino and running the openvino checks if import was successful.

Removed OpenVINO check from there. Next release should fix Gather.

ssheorey reviewed (commented):
Thanks! Almost ready to merge, I think. Can you add a link to `docs/openvino.md` from `readme.md`, and 2 minor comments below.

> Reviewed 1 of 12 files at r1, 4 of 11 files at r5, 1 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/406#-:-MrU-G2v7Mbq9fF0itDn:b-nzrp5t)* status: 6 of 14 files reviewed, 16 unresolved discussions (waiting on @dkurt, @sanskar107, and @ssheorey)

___
*[docs/openvino.md, line 3 at r6](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrTzVYR3H9U2_XeSJqw:-MrTzVYS4q1Kfn3pdq7d:bgb1sps) ([raw file](https://github.com/isl-org/open3d-ml/blob/489b16dc1b07e5923ed77041b715b9393839c23e/docs/openvino.md#L3)):*
> ```Markdown
> # OpenVINO backend
> 
> Open3D-ML allows to use [Intel OpenVINO](https://github.com/openvinotoolkit/openvino) as an optional backend for deep learning models.
> ```

```
Open3D-ML allows to use [Intel OpenVINO](https://github.com/openvinotoolkit/openvino) as an optional backend for deep learning model inference. This allows you to run 3DML model inference on different Intel devices such as CPU, GPU and the Myriad family VPUs.
```
Clarify that this is for inference, not training. [Many Open3D-ML users will be interested in training].
___
*[docs/openvino.md, line 15 at r6](https://reviewable.io/reviews/isl-org/open3d-ml/406#-MrU-8P50NDhrvcTCVoB:-MrU-8P6EJHNuRo4yEJT:bmkui5r) ([raw file](https://github.com/isl-org/open3d-ml/blob/489b16dc1b07e5923ed77041b715b9393839c23e/docs/openvino.md#L15)):*
> ```Markdown
> 
> Then use `net` as usual.
> ```

## Supported models

Add model support matrix.


<!-- Sent from Reviewable.io -->

dkurt commented:
@ssheorey, thanks! Updated.

ssheorey reviewed (approved):
Great, thanks @dkurt !
